Add Teradata connector

sc250072 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information 
at https://trino.io/development/process.html, 
at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md 
and contact us on #core-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description
This PR introduces the Teradata connector, enabling Trino to query and interact with Teradata data sources.

Key additions:

Implements the Teradata connector SPI with support for:
• Connection setup
• Schema retrieval
• Data type mapping

Adds integration tests against a Teradata ClearScape Analytics™ Experience instance to verify connection, query execution, and result mapping.
Updates documentation under connectors to include usage examples and configuration options.
This connector extends Trino’s ecosystem to support Teradata users, allowing them to leverage Trino’s distributed query engine for federated queries.


<!-- Provide details that help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
The Teradata connector module has both unit tests and integration tests.
The integration tests require access to a [Teradata ClearScape Analytics™ Experience](https://clearscape.teradata.com/sign-in).
You can follow the steps below to run the integration tests locally.

> A new PR has been created for the Teradata connector to replace [PR #26574](https://github.com/trinodb/trino/pull/26574)
, with a reduced scope based on the recommendation in [this review comment](https://github.com/trinodb/trino/pull/26574#discussion_r2345746054)
. Additional review comments from the earlier PR have also been addressed.

### Prerequisites
1. Create a new ClearScape Analytics™ Experience account
    
   If you don't already have one, sign up at:

   [Teradata ClearScape Analytics™ Experience](https://www.teradata.com/getting-started/demos/clearscape-analytics)

2. Login

   Sign in with your new account at:

   [ClearScape Analytics™ Experience Login](https://clearscape.teradata.com/sign-in)

3. Collect the API Token

    Use the Copy API Token button in the UI to retrieve your token.

4. Define the following environment variables
 
   ⚠️ Note: The Teradata database password must be at least 8 characters long.

    ```
    export CLEARSCAPE_TOKEN=<API Token>
    export CLEARSCAPE_PASSWORD=<Password for Teradata database (min 8 chars)>
    export CLEARSCAPE_REGION=us-central;
    ```

### Running Integration Tests
Once the environment variables are set, run the integration tests with:

⚠️ Note: Run the following command from the Trino parent directory.

`./mvnw clean install -pl :trino-teradata`


<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Connectors
* Add Teradata connector. ({issue}`<issue-number-if-any>`)
```

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/integration/TeradataConnectorTest.java:
```diff
@@ -0,0 +1,478 @@
+
+    @AfterAll
+    public void cleanupTestDatabase()
+    {
```
updated with closeAfterClass

findepi commented:
> The Teradata connector module has both unit tests and integration tests.
> The integration tests require access to a [Teradata ClearScape Analytics™ Experience](https://clearscape.teradata.com/sign-in).

How will these tests be run on Trino CI?

sc250072 commented on plugin/trino-teradata/pom.xml:
```diff
@@ -0,0 +1,282 @@
+
+        <dependency>
+            <groupId>com.teradata.jdbc</groupId>
+            <artifactId>terajdbc</artifactId>
```
Hi @findepi, Thanks for pointing this out. The Teradata JDBC driver is distributed under the "Teradata Generic Download License" and is free to download and use.

We are currently using the same driver in the Airbyte project for the Teradata connector without issues.

That said, we are checking internally with our legal team to confirm whether this license is compatible for inclusion in Apache projects like Trino. We’ll get back to you as soon as we have a confirmation.

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/integration/TestingTeradataServer.java:
```diff
@@ -0,0 +1,281 @@
+        // Initialize ClearScape Instance and Get the host name from ClearScape API in case config is using clearscape
+        if (config.isUseClearScape()) {
+            boolean destroyEnv = false;
+            if (System.getenv("CLEARSCAPE_DESTORY_ENV") != null) {
```
Updated with io.trino.testing.SystemEnvironmentUtils. Thank you for suggestion

electrum commented:
There is only one `LogonMechanism` type. Are there other types that will be supported in the future? If not, we should remove the configuration.

sc250072 commented:
> I created a new PR here with a number of minor fix ups: #27206
> 
> If you're happy with the changes, we can squash them into your commit

The changes look good to me. I’ve run the tests and everything works as expected. Thanks for making the updates!

ebyhr commented:
@sc250072 Could you rebase on master and remove unrelated commits?

ebyhr commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,178 @@
+source by appending parameters to the JDBC connection string set in the 
+connection-url catalog configuration property.
+
+For example, to specify SSLMODE:
```
```suggestion
For example, to specify `SSLMODE`:
```

sc250072 replied:
Updated

ebyhr commented on docs/src/main/sphinx/connector/teradata.md:
````diff
@@ -0,0 +1,178 @@
+The Teradata connector provides a schema for every Teradata database. You can
+see the available Teradata databases by running SHOW SCHEMAS:
+
+```
````
Same for other SQL examples:
```suggestion
```sql
```

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+        return Optional.empty();
+    }
+
+    private Optional<ColumnMapping> numberMapping(JdbcTypeHandle typeHandle)
```
This method can be static.

sc250072 replied:
changed. Modified other mapping methods - charColumnMapping, varcharColumnMapping

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataConfig.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Config("logon-mechanism")
+    @ConfigDescription("Specifies the logon mechanism for Teradata (default: TD2). Use 'TD2' for TD2 authentication.")
+    public TeradataConfig setLogMech(String logMech)
```
You can use enum instead of String here. Then, you can remove "Unsupported logon mechanism" check from TeradataClientModule.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/unit/TestTeradataPlugin.java:
```diff
@@ -0,0 +1,43 @@
+import io.trino.plugin.teradata.TeradataPlugin;
+import io.trino.spi.connector.ConnectorFactory;
+import io.trino.testing.TestingConnectorContext;
+import org.junit.jupiter.api.Assertions;
```
Use AssertJ assertion instead.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/integration/clearscape/CreateEnvironmentRequest.java:
```diff
@@ -0,0 +1,20 @@
+        String name,
+        String region,
+        String password
+) {}
```
Add compact constructor with `requireNonNull`. Same for other record classes.

ebyhr reviewed (commented):
Left initial comments.

sc250072 commented:
> I created a new PR here with a number of minor fix ups: #27206
> 
> If you're happy with the changes, we can squash them into your commit

@electrum Could you please squash the changes so I can address the new review comments? Also, could you configure the ClearScape secrets on this branch so we can run the Teradata integration tests?

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataQueryRunner.java:
```diff
@@ -0,0 +1,126 @@
+        {
+            @Language("SQL") String sql = String.format("CREATE TABLE %s AS SELECT * FROM %s", table.objectName(), table);
+            queryRunner.execute(session, sql);
+            ((ObjectAssert) assertThat(queryRunner.execute(session, "SELECT count(*) FROM " + table.objectName()).getOnlyValue()).as("Table is not loaded properly: %s", new Object[] {
```
Removed

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataQueryRunner.java:
```diff
@@ -0,0 +1,126 @@
+        return new Builder(server);
+    }
+
+    public static void main(String[] args)
```
Followed same and updated accordingly

sc250072 commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,178 @@
+Refer to the following sections for type mapping in when reading data from
+Teradata to Trino.
+
+### Teradata type to Trino type mapping
```
added

ebyhr commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+# Teradata connector
```
@mosabua Do you want to review this doc?

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+        // try to use result set metadata from select * from table to populate the mapping
+        try {
+            HashMap<String, CaseSensitivity> caseMap = new HashMap<>();
+            String sql = format("select * from %s.%s where 0=1", schemaTableName.getSchemaName(), schemaTableName.getTableName());
```
No need to add `quoted` to the schema and table names?

sc250072 replied:
Thanks for pointing out this. Updated with `quoted` and Uppercase for select, from, where.
```java
 String sql = format("SELECT * FROM %s.%s WHERE 0=1", quoted(schemaTableName.getSchemaName()), quoted(schemaTableName.getTableName()));

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+    @Override
+    public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        // this method should ultimately encompass all the expected teradata data types
```
Remove an helpful comment.

sc250072 replied:
Removed

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+    private static ColumnMapping charColumnMapping(int charLength, boolean isCaseSensitive)
+    {
+        if (charLength > CharType.MAX_LENGTH) {
+            return varcharColumnMapping(charLength, isCaseSensitive);
```
Fallback to varchar type is weird. Could you leave a code comment?

sc250072 replied:
I’ve removed the if loop since it’s not required in this case. Teradata’s maximum CHAR/VARCHAR length (64,000) is always within Trino’s supported CharType.MAX_LENGTH (65,536), so the fallback branch would never be triggered.

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+            case Type typeInstance when typeInstance == DOUBLE -> WriteMapping.doubleMapping("double precision", doubleWriteFunction());
+            case Type typeInstance when typeInstance == DATE -> WriteMapping.longMapping("date", dateWriteFunctionUsingLocalDate());
+            case DecimalType decimalTypeInstance -> {
+                String dataType = String.format("decimal(%s, %s)", decimalTypeInstance.getPrecision(), decimalTypeInstance.getScale());
```
The format method is imported statically.

sc250072 replied:
Used Static import format method

ebyhr commented on plugin/trino-teradata/pom.xml:
```diff
@@ -0,0 +1,346 @@
+        <dependency>
+            <groupId>com.teradata.jdbc</groupId>
+            <artifactId>terajdbc</artifactId>
+            <version>20.00.00.49</version>
```
```suggestion
            <version>20.00.00.51</version>
```

sc250072 replied:
Updated with latest recent release of Teradata Jdbc driver version

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/ClearScapeManager.java:
```diff
@@ -0,0 +1,175 @@
+    public void init(Model model)
+    {
+        this.model = model;
+    }
```
Can we add a constructor with `Model` instead?

sc250072 replied:
Yes. Added constructor and removed init

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/TeradataHttpClient.java:
```diff
@@ -0,0 +1,166 @@
+            String baseUrl)
+    {
+        this.httpClient = httpClient;
+        this.baseUrl = baseUrl;
```
Add `requireNonNull`.

sc250072 replied:
Added requireNonNull

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/TeradataHttpClient.java:
```diff
@@ -0,0 +1,166 @@
+
+    // Creating an environment is a blocking operation by default, and it takes ~1.5min to finish
+    public CompletableFuture<EnvironmentResponse> createEnvironment(CreateEnvironmentRequest createEnvironmentRequest,
+            String token)
```
Unwrap.

sc250072 replied:
Done.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/TeradataHttpClient.java:
```diff
@@ -0,0 +1,166 @@
+    public CompletableFuture<EnvironmentResponse> createEnvironment(CreateEnvironmentRequest createEnvironmentRequest,
+            String token)
+    {
+        var requestBody = handleCheckedException(() -> objectMapper.writeValueAsString(createEnvironmentRequest));
```
Avoid `var` https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-var
Same for others.

sc250072 replied:
Thank you for the reference. Removed var usage across the Teradata connector plugin

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/TeradataHttpClient.java:
```diff
@@ -0,0 +1,166 @@
+        else if (httpResponse.statusCode() >= 500 && httpResponse.statusCode() <= 599) {
+            throw new Error5xxException(httpResponse.statusCode(), body);
+        }
+        else {
```
Redundant else.

sc250072 replied:
Merged Error4xxException and Error5xxException to single exception class `ClearScapeServiceException`.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TestTeradataTypeMapping.java:
```diff
@@ -0,0 +1,282 @@
+                .execute(getQueryRunner(), teradataJDBCCreateAndInsert("date"));
+    }
+
+    private DataSetup teradataJDBCCreateAndInsert(String tableNamePrefix)
```
```suggestion
    private DataSetup teradataCreateAndInsert(String tableNamePrefix)
```

sc250072 replied:
Modified to teradataCreateAndInsert

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+
+To connect to Teradata, you need:
+
+- Teradata Database
```
```suggestion
- Teradata database.
```

Any minimum version ?

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+
+### Multiple Teradata databases
+
+You can have as many catalogs as you need, so if you have additional Teradata
```
```suggestion
You can configure as many catalogs as you need, so if you have additional Teradata
```

sc250072 replied:
Revised `Multiple Teradata databases` section to align with Postgres and SQL Server connector content.

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+You can have as many catalogs as you need, so if you have additional Teradata
+databases, simply add another properties file to etc/catalog with a different
+name, making sure it ends in .properties. 
+For example, if you name the property file sales.properties, Trino creates a 
```
```suggestion
For example, if you name the catalog file `sales.properties`, Trino creates a 
```

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+databases, simply add another properties file to etc/catalog with a different
+name, making sure it ends in .properties. 
+For example, if you name the property file sales.properties, Trino creates a 
+catalog named sales using the configured connector.
```
```suggestion
catalog named `sales` using the configured connector.
```

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
````diff
@@ -0,0 +1,214 @@
+```{include} jdbc-type-mapping.fragment
+```
+
+## Querying Teradata
````
Remove this whole section

sc250072 replied:
Removed

mosabua commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+
+## SQL support
+
+The connector provides read access to data and metadata in
```
This is inconsistent with intro and type mapping .. those sections imply that some DDL statements are suppported.. please correct and use wording/formmatting like in other connectors

sc250072 replied:
Updated with correct formatting and added about supported `DDL` statements.

sc250072 commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+        this.teradataJDBCCaseSensitivity = teradataConfig.getTeradataCaseSensitivity();
+    }
+
+    private static ColumnMapping charColumnMapping(int charLength, boolean isCaseSensitive)
```
Moved helper methods after the callers for teradata to follow project patter

sc250072 commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+    protected void createSchema(ConnectorSession session, Connection connection, String remoteSchemaName)
+    {
+        execute(session, format(
+                "CREATE DATABASE %s AS PERMANENT = 60000000, SPOOL = 120000000",
```
`Teradata` requires the permanent storage value to be explicitly specified in bytes, as no default is provided. We are using 60,000,000 bytes, consistent with the storage value defined in other Teradata connectors used by the tools

sc250072 commented on pom.xml:
```diff
@@ -113,6 +113,7 @@
         <module>plugin/trino-snowflake</module>
         <module>plugin/trino-spooling-filesystem</module>
         <module>plugin/trino-sqlserver</module>
+        <module>plugin/trino-teradata</module>
```
When we add an entry for Teradata in EnvMultinodeAllConnectors, the trino-product-tests-launcher plugin build fails with an unhandled exception reported by the Error Prone static analysis plugin at EnvMultinodeAllConnectors.java:[78,25]

sc250072 replied:
@ebyhr  added Teradata to EnvMultinodeAllConnectors and introduced a TestTeradata product test to verify class loader isolation. Could you please review and let me know if any adjustments are needed?

sc250072 commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+## Configuration
+
+To configure the Teradata connector, create a catalog properties file in
+`etc/catalog` named, for example, `example.properties`, to mount the Teradata
```
Revised `Configuration` section to align with Postgres connector content.

sc250072 commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,214 @@
+
+- Teradata Database
+- Network access from the Trino coordinator and workers to Teradata. Port 
+  1025 is the default port
```
Done

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataTestConstants.java:
```diff
@@ -0,0 +1,20 @@
+{
+    String ENV_CLEARSCAPE_URL = "https://api.clearscape.teradata.com";
+    String ENV_CLEARSCAPE_USERNAME = "demo_user";
+}
```
Applied the suggested changes to `TeradataTestConstants`: converted to a final class and added the final keyword to both CLEARSCAPE_URL and CLEARSCAPE_USERNAME.

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataQueryRunner.java:
```diff
@@ -0,0 +1,124 @@
+            logger.setLevel("io.trino.plugin.teradata", Level.DEBUG);
+            logger.setLevel("io.trino", Level.INFO);
+            TestingTeradataServer server = new TestingTeradataServer("TeradataQueryRunner", false);
+            QueryRunner queryRunner = builder(server).addCoordinatorProperty("http-server.http.port", "8080").setInitialTables(TpchTable.getTables()).build();
```
Wrapped lines and applied the suggestion across all relevant sections.

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataQueryRunner.java:
```diff
@@ -0,0 +1,124 @@
+            return super.build();
+        }
+
+        public static void main(String[] args)
```
Updated

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/TeradataHttpClient.java:
```diff
@@ -0,0 +1,166 @@
+
+    public TeradataHttpClient(
+            HttpClient httpClient,
+            String baseUrl)
```
Removed redundant constructor. 

```java
public TeradataHttpClient(String baseUrl) {
    requireNonNull(baseUrl, "baseUrl should not be null");
    this.baseUrl = baseUrl;
    httpClient = HttpClient.newBuilder()
            .version(HttpClient.Version.HTTP_1_1)
            .build();
    objectMapper = JsonMapper.builder()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .configure(MapperFeature.ALLOW_FINAL_FIELDS_AS_MUTATORS, false)
            .build();
}

sc250072 commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/ClearScapeManager.java:
```diff
@@ -0,0 +1,175 @@
+        if (isValidUrl(envUrl)) {
+            return new TeradataHttpClient(envUrl);
+        }
+        else {
```
Removed.

sc250072 commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,382 @@
+            for (int i = 1; i <= columnCount; i++) {
+                caseMap.put(rsmd.getColumnName(i), rsmd.isCaseSensitive(i) ? CASE_SENSITIVE : CASE_INSENSITIVE);
+            }
+            pstmt.close();
```
Updated block with  try-with-resources.
```java
try (PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

sc250072 commented:
@ebyhr @mosabua All review comments have been addressed. Could you please review the updated changes?

mosabua reviewed (commented):
Docs look good now .. beyond that I leave it to @ebyhr @electrum and others.

crispkid commented:
I’ve rebuilt the Trino Teradata Connector for version 479.
[https://github.com/crispkid/trino-teradata](url)

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,352 @@
+                isCaseSensitive ? FULL_PUSHDOWN : CASE_INSENSITIVE_CHARACTER_PUSHDOWN);
+    }
+
+    private boolean deriveCaseSensitivity(CaseSensitivity caseSensitivity)
```
This method can be static.

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClientModule.java:
```diff
@@ -0,0 +1,62 @@
+        connectionProperties.setProperty("LOGMECH", "TD2");
+        return DriverConnectionFactory.builder(driver, config.getConnectionUrl(), credentialProvider)
+                .setConnectionProperties(connectionProperties)
+                .setOpenTelemetry(openTelemetry).build();
```
What happens if connection URL contains a different LOGMECH?

ebyhr commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClientModule.java:
```diff
@@ -0,0 +1,62 @@
+    public void setup(Binder binder)
+    {
+        binder.bind(JdbcClient.class).annotatedWith(ForBaseJdbc.class).to(TeradataClient.class).in(Scopes.SINGLETON);
+        configBinder(binder).bindConfig(JdbcStatisticsConfig.class);
```
Does this connector support table statistics?

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/ClearScapeManager.java:
```diff
@@ -0,0 +1,173 @@
+    public ClearScapeManager(Model model)
+    {
+        requireNonNull(model, "model should not be null");
+        this.model = model;
```
We usually use `xxx is null` message. Same for others. 
```suggestion
        this.model = requireNonNull(model, "model is null");
```

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/ClearScapeSetup.java:
```diff
@@ -0,0 +1,91 @@
+        this.password = password;
+        this.envName = envName;
+        this.region = region;
+        this.destroyEnv = destroyEnv;
```
There is no need to separate requireNonNull and field initialization. Same for other places.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/GetEnvironmentRequest.java:
```diff
@@ -0,0 +1,25 @@
+import static java.util.Objects.requireNonNull;
+
+public record GetEnvironmentRequest(
+        String name)
```
Unwrap.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/clearscape/OperationRequest.java:
```diff
@@ -0,0 +1,25 @@
+import static java.util.Objects.requireNonNull;
+
+public record OperationRequest(
+        String operation)
```
Unwrap.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/DatabaseConfig.java:
```diff
@@ -0,0 +1,156 @@
+    private final AuthenticationConfig authConfig;
+    private final String clearScapeEnvName;
+    private final Map<String, String> jdbcProperties;
+
```
Add a constructor with requireNonNull check.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TeradataQueryRunner.java:
```diff
@@ -0,0 +1,130 @@
+            if (!server.isTableExists(sourceTable)) {
+                copyTable(queryRunner, table, session);
+            }
+        }
```
These methods can be private.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TestTeradataConnectorTest.java:
```diff
@@ -0,0 +1,600 @@
+    @Test
+    public void testDataMappingSmokeTest()
+    {
+        skipTestUnless(false);
```
Why?

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/integration/TestTeradataConnectorTest.java:
```diff
@@ -0,0 +1,600 @@
+    }
+
+    @Override
+    // Override because the expected error message is different
```
The base error message is empty. Remove unhelpful comment.

ebyhr commented on plugin/trino-teradata/src/test/java/io/trino/plugin/teradata/TestTeradataPlugin.java:
```diff
@@ -0,0 +1,40 @@
+public class TestTeradataPlugin
+{
+    @Test
+    public void testCreateConnector()
```
```suggestion
    void testCreateConnector()
```

ebyhr commented on plugin/trino-teradata/pom.xml:
```diff
@@ -0,0 +1,327 @@
+    <description>Trino - Teradata connector</description>
+
+    <properties>
+        <air.check.skip-dependency-scope>true</air.check.skip-dependency-scope>
```
Why do enable this check?

ebyhr commented on testing/trino-product-tests/pom.xml:
```diff
@@ -264,6 +264,13 @@
+        <dependency>
+            <groupId>com.teradata.jdbc</groupId>
+            <artifactId>terajdbc</artifactId>
+            <version>20.00.00.51</version>
```
Define this dependency in root pom.xml so we can avoid repeating the version in several places.

sc250072 commented on docs/src/main/sphinx/connector/teradata.md:
````diff
@@ -0,0 +1,192 @@
+# Teradata connector
+
+```{raw} html
+<img src="../_static/img/teradata.png" class="connector-logo">
````
Changed the background.

sc250072 commented on docs/src/main/sphinx/connector/teradata.md:
```diff
@@ -0,0 +1,192 @@
+
+* - Trino type
+  - Teradata type
+  - Notes
```
Removed

sc250072 commented on plugin/trino-teradata/src/main/java/io/trino/plugin/teradata/TeradataClient.java:
```diff
@@ -0,0 +1,352 @@
+    @Override
+    protected void createSchema(ConnectorSession session, Connection connection, String remoteSchemaName)
+    {
+        execute(session, format("CREATE DATABASE %s AS PERMANENT = 60000000", quoted(remoteSchemaName)));
```
Added new section in `teradata.md` https://github.com/Teradata/trino/blob/teradata-connector/docs/src/main/sphinx/connector/teradata.md#specific-configuration-properties
@mosabua  could you please review this doc change.

mosabua reviewed (approved):
I had another look at the docs. Looks good now.

sc250072 committed:
Add Teradata connector

Implement Teradata Trino Connector to handle Trino operations on Teradata,
including connector configuration, type mappings, ClearScape integration
for test environments, product tests, and documentation.

sc250072 committed:
Add Teradata connector

Implement Teradata Trino Connector to handle Trino operations on Teradata,
including connector configuration, type mappings, ClearScape integration
for test environments, product tests, and documentation.
